### PR TITLE
8237602: TabPane doesn't respect order of TabPane.getTabs() list

### PR DESCRIPTION
**Pull request by arapte**
Issue:
When tabs are permuted as mentioned in the issue description as, 
1. TabPane.getTabs().setAll(tab0, tab1)
2. TabPane.getTabs().setAll(tab0, tab1, tab2, tab3);
the tab headers do not get permuted in same order as `TabPane.getTabs()`.

=> tab headers should be shown in order as  tab0, tab1, tab2, tab3.
=> but are show in order as  tab2, tab3, tab0, tab1

Cause:
Newly added tabs(tab2, tab3) are not inserted at correct index. The index `Change.getFrom()` (0) used from Change does not remain valid after the tabs to be moved(tab0, tab1) are removed from `tabsToAdd` list.

Fix:
Use the index of first newly added tab, from `TabPane.getTabs()`, which would be always reliable.

Verification:
No existing tests fail due to this change.
Added a system test which fails without and pass with fix.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8237602](https://bugs.openjdk.java.net/browse/JDK-8237602): TabPane doesn't respect order of TabPane.getTabs() list

### Reviewers
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)
 * Jeanette Winzenburg ([fastegal](@kleopatra) - Author)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/201/head:pull/201`
`$ git checkout pull/201`

### CODE REVIEW

**Comment by kleopatra** [comment]
didn't look closely, just wondering why you need a system test for this?

**Comment by kevinrushforth** [comment]
The following is wrong in your Description:

> => tab headers should be shown in order as tab2, tab3, tab0, tab1.
> => But should are show in order as tab0, tab1, tab2, tab3

Based on what I see in the bug report (and what I would expect), I'm pretty sure you meant to say:

The tab headers should be shown in order as tab0, tab1, tab2, tab3.
But are shown in order as tab2, tab3, tab0, tab1.

**Comment by arapte** [comment]
> 
> 
> didn't look closely, just wondering why you need a system test for this?

The system test file `TabPanePermuteGetTabsTest` was added to address exact similar kind of issue [JDK-8222457](https://bugs.openjdk.java.net/browse/JDK-8222457).  The current issue [JDK-8237602](https://bugs.openjdk.java.net/browse/JDK-8237602) is just another case of same type. So it seems good to add a new tests with similar existing tests.
I myself had fixed the previous issue, Can't remember why did I not consider of a unit test then 🤔.

**Comment by arapte** [comment]
> 
> 
> The following is wrong in your Description:

Thanks Kevin, the description is corrected now. 👍

**Comment by kleopatra** [comment]
> The system test file `TabPanePermuteGetTabsTest` was added to address exact similar kind of issue [JDK-8222457](https://bugs.openjdk.java.net/browse/JDK-8222457). The current issue [JDK-8237602](https://bugs.openjdk.java.net/browse/JDK-8237602) is just another case of same type. So it seems good to add a new test with similar existing tests.
> I myself had fixed the previous issue, Can't remember why did I not consider of a unit test then 🤔.

yeah, if we don't document everything :))) Would you consider moving both into a unit test? Have one locally, could send it to you .. 

Also a minor quirk as to the naming: it should not name it "permutation" because it isn't from the perspective of the list change: the notification is a "replace". Btw, the pattern in the listener (of collecting up additions/removals) is a bit smelly - will fail as soon as you get disjoint additions (on the bright side: they don't happen in the current implementation of ObservableList but are allowed to), just something to keep in mind :)

**Comment by arapte** [comment]
> Would you consider moving both into a unit test? Have one locally, could send it to you ..

A ready test 👍 , That's very nice of you. Yes, We can move all the four tests to unit test. But I would prefer to do it as another TEST_BUG.  Does that sound good ?

> Btw, the pattern in the listener (of collecting up additions/removals) is a bit smelly

I did try to change whole implementation, but it looked like a big change for the issue we have. The fix is very minimal and local.


> it should not name it "permutation"

Shall think of a new name when we move the tests..

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java [hunk 1]
```diff
@@ -647,7 +645,9 @@ private void initializeTabListener() {
                     }
                 }
 
-                addTabs(tabsToAdd, insertPos == -1 ? tabContentRegions.size() : insertPos);
+                if (!tabsToAdd.isEmpty()) {
+                    addTabs(tabsToAdd, getSkinnable().getTabs().indexOf(tabsToAdd.get(0)));
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:649]: @kleopatra is right in that there is no guarantee that the set of added sublists from multiple Changes are contiguous. Since this assumption is preexisting, I agree that this is just something to keep in mind.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java [hunk 2]
```diff
@@ -991,8 +991,10 @@ private void removeTab(Tab tab) {
         }
 
         private void moveTab(int moveToIndex, TabHeaderSkin tabHeaderSkin) {
-            headersRegion.getChildren().remove(tabHeaderSkin);
-            headersRegion.getChildren().add(moveToIndex, tabHeaderSkin);
+            if (moveToIndex != headersRegion.getChildren().indexOf(tabHeaderSkin)) {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:994]: Unless I am missing something, this check seems unrelated to the bug.

File: tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java [hunk 3]
```diff
@@ -181,6 +189,59 @@ public void testPermuteGetTabsWithMoreTabs() {
         }
     }
 
+    // Test for JDK-8237602
+    @Test
+    public void testPermutGetTabsWithMoreTabs1() {
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java:194]: Typo: Permut --> Permute

(or, as long as you are changing it anyway, you could rename it now to something other than "permute" since it isn't).

File: tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java [hunk 4]
```diff
@@ -181,6 +189,59 @@ public void testPermuteGetTabsWithMoreTabs() {
...
+        // Step #3.2
+        // Click on first tab header
+        selectionLatch[0] = new CountDownLatch(1);
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (scene.getWindow().getX() + scene.getX() + firstTabdXY),
+                    (int) (scene.getWindow().getY() + scene.getY() + firstTabdXY));
+            robot.mousePress(MouseButton.PRIMARY);
+            robot.mouseRelease(MouseButton.PRIMARY);
+        });
+        delay();
```
Comment (kevinrushforth) [inline tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java:224]: Does using Robot to select the Tab test add value? Probably not worth worrying about given that this will all move to the controls unit tests with [JDK-8244195](https://bugs.openjdk.java.net/browse/JDK-8244195).

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java (hunk 2 above)
Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:994]: Yes Kevin, this is not required for fix. In some scenarios the tabHeader gets placed at correct index due to the addition of new headers, so this movement becomes redundant.

File: tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java (hunk 4 above)
Comment (arapte) [reply tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java:224]: In these scenarios, the `TabPane.getTabs()` has correct sequence. So selecting a tab using `TabPaneSelectionModel` does not test the issue. Using Mouse guarantees that `TabHeaderArea` has correct order of tabs. But anyway once we convert these tests to unit tests, mouse events will be removed.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java (hunk 2 above)
Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:994]: In that case, this change is fine.

File: tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java (hunk 3 above)
Comment (arapte) [reply tests/system/src/test/java/test/robot/javafx/scene/TabPanePermuteGetTabsTest.java]: Changed the test name to `testAddingNewTabsWithExistingTabsAtSameIndex`. While converting to unit tests, we will have to change the names of other test from this file.
